### ISSUE TITLE
Use openapi cli plugin to automate swagger api annotation

### ISSUE THREAD

**Issue by nitrosx** [SciCatProject/backend#1590]
## Summary
As highlighted by @emigun, maintenance and consistency of the swagger api documentation is expensive and prone to human error as we need to enter the same information in multiple location.

## Solution
We propose to adopt the [OpenAPI CLI plugin](https://docs.nestjs.com/openapi/cli-plugin), which reduce information duplication, code cluttering and inconsistencies, and allowing complete freedom in configuration where needed.

##Additional information
Below I included an edited version of the [post](https://github.com/SciCatProject/scicat-backend-next/pull/1579#issuecomment-2556686282) reported by @emigun in PR #1579

**Comment by nitrosx** [comment]
Perhaps we should consider using https://docs.nestjs.com/openapi/cli-plugin instead of annotating manually?
It is already enabled.
I have tested it a bit and it seems to work nicely.

A fairly advanced example:
Let's change
```
  @ApiProperty({
    type: "array",
    items: { $ref: getSchemaPath(TechniqueClass) },
    required: false,
    default: [],
    description: "Stores the metadata information for techniques.",
  })
  @IsOptional()
  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => CreateTechniqueDto)
  readonly techniques?: TechniqueClass[] = [];
```

to (remove ApiProperty)

```
  @IsOptional()
  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => CreateTechniqueDto)
  readonly techniques?: TechniqueClass[] = [];
```

That would change the openapi json from:
```
"techniques": {
  "type": "array",
  "items": {
    "$ref": "#/components/schemas/TechniqueClass"
  },
  "default": [],
  "description": "Stores the metadata information for techniques."
},
```

to

```
"techniques": {
  "default": [],
  "type": "array",
  "items": {
    "$ref": "#/components/schemas/TechniqueClass"
  }
}
```

As you can see everything was generated (except description, see below)

I have tried with arrays, dates and ordinary string/number and it seems to work. It also reads the `?` for optional fields and correctly identifies the required ones.

We could change nest-cli.json to:
```{
  "collection": "@nestjs/schematics",
  "sourceRoot": "src",
  "compilerOptions": {
    "plugins": [{
      "name": "@nestjs/swagger",
      "options": {
        "introspectComments": true
      }
    }]
  }
}
```
in order to generate the field "description" based on a comment above. Or you could still use
```
@ApiProperty({
    description: "This is a test description"
  })
```
if you prefer that. The other fields like type, items, default and others will still be generated with the plugin.

tldr; We could probably get rid of all the ApiProperty decorators and the openapi spec will automatically be generated correctly for the fields. If some doesn't work, we can still use ApiProperty on only those. 

Two main benefits:
1. Less cluttered code
2. No more inconstencies (as you can see in my merge request I had to change about 20 things just for the dataset dtos)

**Comment by sbliven** [comment]
Can we drop the type decorators (`@Optional()`, `@IsString()`, etc) as well? I did a quick test and it looks like just having the typescript types is enough, at least for the openapi generation. Are they important for some nest feature?

**Comment by emigun** [comment]
> Can we drop the type decorators (`@Optional()`, `@IsString()`, etc) as well? I did a quick test and it looks like just having the typescript types is enough, at least for the openapi generation. Are they important for some nest feature?

No, they shouldn't be dropped. And yes, they are important for validation in Nest.js. We use a global pipe for validation on dtos.

**Comment by emigun** [comment]
> Can we drop the type decorators (`@Optional()`, `@IsString()`, etc) as well? I did a quick test and it looks like just having the typescript types is enough, at least for the openapi generation. Are they important for some nest feature?

https://github.com/SciCatProject/scicat-backend-next/issues/1590#issuecomment-2735594534 Yes they are important for validation.
